File: julia/AIInAction/src/ch138_fastica.jl
"""
    Ch138Fastica

FastICA for Independent Component Analysis (from scratch, Julia).

Mirrors the Python module `aiinaction.ch138_fastica` and the Rust module
`aiinaction::ch138_fastica`. The shared fixtures in `test/test_ch138_fastica.jl`
match the Python/Rust suites, which keeps the three implementations at parity.

The estimator centers the data, whitens it via the covariance eigendecomposition,
then runs the Hyvarinen-Oja FastICA fixed-point iteration with the `logcosh`
contrast (`g(u) = tanh(u)`) and symmetric orthogonalization
`W <- (W W^T)^{-1/2} W`. For cross-language determinism the unmixing matrix is
initialized to the identity, a fixed iteration count is used (no tolerance stop),
and a self-contained cyclic Jacobi eigensolver (the same one used by the PCA
chapter) drives both whitening and orthogonalization. Components are ordered by
descending recovered-source variance and each unmixing row's largest-magnitude
entry is forced positive.
"""
module Ch138Fastica

export ICAResult, fit_ica, transform

"""The fitted state of a FastICA model. `components` are rows (one source per row)."""
struct ICAResult
    mean::Vector{Float64}
    whitening::Matrix{Float64}
    unmixing::Matrix{Float64}
    components::Matrix{Float64}
    mixing::Matrix{Float64}
    n_iter::Int
end

n_components(r::ICAResult) = size(r.components, 1)
n_features(r::ICAResult) = size(r.components, 2)

function _as_matrix(X)
    A = Matrix{Float64}(X)
    n, d = size(A)
    n >= 2 || throw(ArgumentError("need at least 2 samples, got $n"))
    d >= 1 || throw(ArgumentError("X must have at least one feature"))
    all(isfinite, A) || throw(ArgumentError("X contains non-finite values (nan or inf)"))
    return A
end

"""Symmetric eigendecomposition by the cyclic Jacobi method.

Returns `(eigenvalues, V)` sorted by descending eigenvalue, where eigenvector `k`
is column `k` of `V`."""
function _jacobi_eigen(A0::Matrix{Float64})
    n = size(A0, 1)
    A = copy(A0)
    V = Matrix{Float64}(undef, n, n)
    fill!(V, 0.0)
    @inbounds for i in 1:n
        V[i, i] = 1.0
    end
    for _ in 1:100
        off = 0.0
        @inbounds for p in 1:n, q in (p + 1):n
            off += A[p, q]^2
        end
        off < 1e-30 && break
        @inbounds for p in 1:n, q in (p + 1):n
            apq = A[p, q]
            abs(apq) < 1e-300 && continue
            theta = (A[q, q] - A[p, p]) / (2 * apq)
            t = theta == 0 ? 1.0 : sign(theta) / (abs(theta) + sqrt(theta^2 + 1))
            c = 1 / sqrt(t^2 + 1)
            s = t * c
            for k in 1:n
                akp = A[k, p]
                akq = A[k, q]
                A[k, p] = c * akp - s * akq
                A[k, q] = s * akp + c * akq
            end
            for k in 1:n
                apk = A[p, k]
                aqk = A[q, k]
                A[p, k] = c * apk - s * aqk
                A[q, k] = s * apk + c * aqk
            end
            for k in 1:n
                vkp = V[k, p]
                vkq = V[k, q]
                V[k, p] = c * vkp - s * vkq
                V[k, q] = s * vkp + c * vkq
            end
        end
    end
    eigvals = [A[i, i] for i in 1:n]
    order = sortperm(eigvals; rev=true)
    return eigvals[order], V[:, order]
end

"""Symmetric orthogonalization `W <- (W W^T)^{-1/2} W` via Jacobi eigen."""
function _symmetric_decorrelation(W::Matrix{Float64})
    eigvals, V = _jacobi_eigen(W * W')
    inv_sqrt = V * Diagonal_like(1.0 ./ sqrt.(eigvals)) * V'
    return inv_sqrt * W
end

# Build a dense diagonal matrix without depending on LinearAlgebra.Diagonal.
function Diagonal_like(v::Vector{Float64})
    n = length(v)
    M = zeros(Float64, n, n)
    @inbounds for i in 1:n
        M[i, i] = v[i]
    end
    return M
end

function _fix_signs!(W::AbstractMatrix{Float64})
    for j in 1:size(W, 1)
        row = @view W[j, :]
        k = argmax(abs.(row))
        if row[k] < 0
            row .= .-row
        end
    end
    return W
end

"""Moore-Penrose pseudoinverse of a `k x d` matrix via the normal equations.

For the square, full-rank operators produced here this is the exact inverse:
`A^+ = A^T (A A^T)^{-1}`, returning a `d x k` matrix."""
function _pseudoinverse(A::Matrix{Float64})
    k = size(A, 1)
    G = A * A'
    eigvals, V = _jacobi_eigen(G)
    Ginv = V * Diagonal_like(1.0 ./ eigvals) * V'
    return A' * Ginv
end

"""
    fit_ica(X; n_components=nothing, max_iter=200)

Fit a FastICA model to the `n x d` matrix `X` (samples in rows, signals in
columns). Defaults to `d` components. `max_iter` is a fixed iteration count (no
tolerance stop) so the result is deterministic across languages.
"""
function fit_ica(X; n_components=nothing, max_iter::Integer=200)
    A = _as_matrix(X)
    n, d = size(A)

    k = n_components === nothing ? d : Int(n_components)
    (1 <= k <= d) ||
        throw(ArgumentError("n_components must be in [1, $d] for a $(n)x$(d) matrix, got $k"))
    max_iter >= 1 || throw(ArgumentError("max_iter must be a positive integer, got $max_iter"))

    # 1. Center.
    mu = vec(sum(A; dims=1)) ./ n
    Xc = A .- mu'

    # 2. Whiten. Cov = Xc' Xc / (n - 1), d x d.
    cov = (Xc' * Xc) ./ (n - 1)
    eigvals, E = _jacobi_eigen(Matrix(cov))
    any(eigvals[1:k] .<= 0.0) &&
        throw(ArgumentError("data is rank-deficient: a retained whitening direction has zero variance"))
    Ek = E[:, 1:k]
    # K maps centered data to whitened: z = K x_c, shape k x d.
    K = (Ek ./ sqrt.(eigvals[1:k])')'
    K = Matrix{Float64}(K)
    Z = Xc * K'  # n x k

    # 3. FastICA fixed-point iteration with symmetric orthogonalization.
    W = _symmetric_decorrelation(Matrix{Float64}(I_like(k)))
    n_iter = 0
    for _ in 1:max_iter
        n_iter += 1
        WZ = Z * W'              # n x k, column j is w_j^T z
        gwz = tanh.(WZ)
        gprime = 1.0 .- gwz .^ 2
        # w_j^+ = E[z g(w^T z)] - E[g'(w^T z)] w_j
        Wnew = (gwz' * Z) ./ n .- (vec(sum(gprime; dims=1) ./ n) .* W)
        W = _symmetric_decorrelation(Matrix{Float64}(Wnew))
    end
    _fix_signs!(W)

    components = W * K  # k x d

    # Order by descending recovered-source variance.
    S = Xc * components'  # n x k
    Smu = vec(sum(S; dims=1) ./ n)
    var = vec(sum(abs2, S .- Smu'; dims=1) ./ (n - 1))
    order = sortperm(var; rev=true)
    W = W[order, :]
    components = components[order, :]

    mixing = _pseudoinverse(Matrix{Float64}(components))

    return ICAResult(mu, K, Matrix{Float64}(W), Matrix{Float64}(components),
        Matrix{Float64}(mixing), n_iter)
end

# Identity matrix without depending on LinearAlgebra.I.
function I_like(n::Int)
    M = zeros(Float64, n, n)
    @inbounds for i in 1:n
        M[i, i] = 1.0
    end
    return M
end

"""Recover the independent sources from observed mixtures `X`, returning `n x k`."""
function transform(model::ICAResult, X)
    A = Matrix{Float64}(X)
    size(A, 2) == n_features(model) ||
        throw(ArgumentError("X has $(size(A, 2)) features but model was fit on $(n_features(model))"))
    all(isfinite, A) || throw(ArgumentError("X contains non-finite values (nan or inf)"))

    Xc = A .- model.mean'
    return Xc * model.components'
end

end # module Ch138Fastica

File: rust/aiinaction/src/ch138_fastica.rs
//! FastICA for Independent Component Analysis (from scratch, Rust).
//!
//! Mirrors the Python module `aiinaction.ch138_fastica` and the Julia module
//! `AIInAction.Ch138Fastica`. The shared fixtures in the tests below match the
//! Python/Julia suites, which is what keeps the three implementations at parity.
//!
//! This is a std-only implementation of the Hyvarinen-Oja FastICA fixed-point
//! algorithm with the `logcosh` contrast (`g(u) = tanh(u)`). The pipeline centers
//! the data, whitens it with the covariance eigendecomposition, then runs FastICA
//! with **symmetric** orthogonalization `W <- (W W^T)^{-1/2} W`.
//!
//! Determinism for cross-language parity: the unmixing matrix is initialized to
//! the identity, a fixed number of iterations is run (no tolerance stop), and both
//! the whitening and the symmetric orthogonalization use the same self-contained
//! cyclic Jacobi eigensolver as the PCA chapter. Components are ordered by
//! descending recovered-source variance and each unmixing row's largest-magnitude
//! entry is forced positive.

/// A dense row-major matrix of `f64`.
#[derive(Clone, Debug)]
pub struct Matrix {
    pub rows: usize,
    pub cols: usize,
    pub data: Vec<f64>,
}

impl Matrix {
    /// Builds a matrix from a slice of equal-length rows.
    pub fn from_rows(rows: &[Vec<f64>]) -> Result<Matrix, String> {
        if rows.is_empty() {
            return Err("X must have at least one row".to_string());
        }
        let cols = rows[0].len();
        if cols == 0 {
            return Err("X must have at least one feature".to_string());
        }
        let mut data = Vec::with_capacity(rows.len() * cols);
        for r in rows {
            if r.len() != cols {
                return Err("all rows must have the same length".to_string());
            }
            data.extend_from_slice(r);
        }
        Ok(Matrix {
            rows: rows.len(),
            cols,
            data,
        })
    }

    #[inline]
    fn get(&self, i: usize, j: usize) -> f64 {
        self.data[i * self.cols + j]
    }
}

/// The fitted state of a FastICA model.
#[derive(Clone, Debug)]
pub struct IcaResult {
    /// Per-signal training means, length `d`.
    pub mean: Vec<f64>,
    /// Whitening matrix `K`, shape `n_components x d`: `z = K x_c`.
    pub whitening: Vec<Vec<f64>>,
    /// Orthogonal unmixing rotation `W`, shape `n_components x n_components`.
    pub unmixing: Vec<Vec<f64>>,
    /// Full unmixing operator `W K`, shape `n_components x d`.
    pub components: Vec<Vec<f64>>,
    /// Estimated mixing matrix (pseudoinverse of `components`), shape `d x n_components`.
    pub mixing: Vec<Vec<f64>>,
    /// Number of fixed-point iterations run.
    pub n_iter: usize,
}

impl IcaResult {
    pub fn n_components(&self) -> usize {
        self.components.len()
    }
    pub fn n_features(&self) -> usize {
        self.mean.len()
    }
}

/// Symmetric eigendecomposition via the cyclic Jacobi method.
///
/// Returns `(eigenvalues, eigenvectors)` where eigenvector `k` is column `k` of the
/// returned matrix (stored row-major, `n x n`), sorted by descending eigenvalue.
fn jacobi_eigen(a_in: &[f64], n: usize) -> (Vec<f64>, Vec<f64>) {
    let mut a = a_in.to_vec();
    let mut v = vec![0.0f64; n * n];
    for i in 0..n {
        v[i * n + i] = 1.0;
    }
    let idx = |i: usize, j: usize| i * n + j;

    for _sweep in 0..100 {
        let mut off = 0.0;
        for p in 0..n {
            for q in (p + 1)..n {
                off += a[idx(p, q)] * a[idx(p, q)];
            }
        }
        if off < 1e-30 {
            break;
        }
        for p in 0..n {
            for q in (p + 1)..n {
                let apq = a[idx(p, q)];
                if apq.abs() < 1e-300 {
                    continue;
                }
                let app = a[idx(p, p)];
                let aqq = a[idx(q, q)];
                let theta = (aqq - app) / (2.0 * apq);
                let t = if theta == 0.0 {
                    1.0
                } else {
                    theta.signum() / (theta.abs() + (theta * theta + 1.0).sqrt())
                };
                let c = 1.0 / (t * t + 1.0).sqrt();
                let s = t * c;
                for k in 0..n {
                    let akp = a[idx(k, p)];
                    let akq = a[idx(k, q)];
                    a[idx(k, p)] = c * akp - s * akq;
                    a[idx(k, q)] = s * akp + c * akq;
                }
                for k in 0..n {
                    let apk = a[idx(p, k)];
                    let aqk = a[idx(q, k)];
                    a[idx(p, k)] = c * apk - s * aqk;
                    a[idx(q, k)] = s * apk + c * aqk;
                }
                for k in 0..n {
                    let vkp = v[idx(k, p)];
                    let vkq = v[idx(k, q)];
                    v[idx(k, p)] = c * vkp - s * vkq;
                    v[idx(k, q)] = s * vkp + c * vkq;
                }
            }
        }
    }

    let eigvals: Vec<f64> = (0..n).map(|i| a[idx(i, i)]).collect();
    let mut order: Vec<usize> = (0..n).collect();
    order.sort_by(|&x, &y| eigvals[y].partial_cmp(&eigvals[x]).unwrap());

    let sorted_vals: Vec<f64> = order.iter().map(|&k| eigvals[k]).collect();
    let mut sorted_vecs = vec![0.0f64; n * n];
    for (k, &col) in order.iter().enumerate() {
        for i in 0..n {
            sorted_vecs[idx(i, k)] = v[idx(i, col)];
        }
    }
    (sorted_vals, sorted_vecs)
}

/// Symmetric orthogonalization `W <- (W W^T)^{-1/2} W` for a `k x k` matrix `w`.
fn symmetric_decorrelation(w: &[Vec<f64>], k: usize) -> Vec<Vec<f64>> {
    // M = W W^T, k x k.
    let mut m = vec![0.0f64; k * k];
    for i in 0..k {
        for j in 0..k {
            let mut acc = 0.0;
            for t in 0..k {
                acc += w[i][t] * w[j][t];
            }
            m[i * k + j] = acc;
        }
    }
    let (eigvals, eigvecs) = jacobi_eigen(&m, k);
    // inv_sqrt = V diag(1/sqrt(lambda)) V^T.
    let mut inv_sqrt = vec![0.0f64; k * k];
    for i in 0..k {
        for j in 0..k {
            let mut acc = 0.0;
            for c in 0..k {
                acc += eigvecs[i * k + c] * (1.0 / eigvals[c].sqrt()) * eigvecs[j * k + c];
            }
            inv_sqrt[i * k + j] = acc;
        }
    }
    // out = inv_sqrt * W.
    let mut out = vec![vec![0.0f64; k]; k];
    for i in 0..k {
        for j in 0..k {
            let mut acc = 0.0;
            for c in 0..k {
                acc += inv_sqrt[i * k + c] * w[c][j];
            }
            out[i][j] = acc;
        }
    }
    out
}

fn fix_signs(w: &mut [Vec<f64>]) {
    for row in w.iter_mut() {
        let mut k = 0usize;
        let mut best = 0.0f64;
        for (j, &val) in row.iter().enumerate() {
            if val.abs() > best {
                best = val.abs();
                k = j;
            }
        }
        if row[k] < 0.0 {
            for val in row.iter_mut() {
                *val = -*val;
            }
        }
    }
}

/// Moore-Penrose pseudoinverse of a `k x d` matrix via the normal equations.
///
/// For the square, full-rank operators produced here this is the exact inverse.
/// Computes `A^+ = A^T (A A^T)^{-1}`, returning a `d x k` matrix.
fn pseudoinverse(a: &[Vec<f64>], k: usize, d: usize) -> Vec<Vec<f64>> {
    // G = A A^T, k x k.
    let mut g = vec![0.0f64; k * k];
    for i in 0..k {
        for j in 0..k {
            let mut acc = 0.0;
            for t in 0..d {
                acc += a[i][t] * a[j][t];
            }
            g[i * k + j] = acc;
        }
    }
    let g_inv = invert_spd(&g, k);
    // A^+ = A^T G^{-1}, d x k.
    let mut out = vec![vec![0.0f64; k]; d];
    for i in 0..d {
        for j in 0..k {
            let mut acc = 0.0;
            for c in 0..k {
                acc += a[c][i] * g_inv[c * k + j];
            }
            out[i][j] = acc;
        }
    }
    out
}

/// Inverts a symmetric positive-definite `n x n` matrix via Jacobi eigen.
fn invert_spd(m: &[f64], n: usize) -> Vec<f64> {
    let (eigvals, eigvecs) = jacobi_eigen(m, n);
    let mut inv = vec![0.0f64; n * n];
    for i in 0..n {
        for j in 0..n {
            let mut acc = 0.0;
            for c in 0..n {
                acc += eigvecs[i * n + c] * (1.0 / eigvals[c]) * eigvecs[j * n + c];
            }
            inv[i * n + j] = acc;
        }
    }
    inv
}

/// Fits a FastICA model to `x` (samples in rows, signals in columns).
///
/// `n_components = None` recovers `d` components. `max_iter` is a fixed iteration
/// count (no tolerance stop) so the result is deterministic across languages.
pub fn fit_ica(
    x: &Matrix,
    n_components: Option<usize>,
    max_iter: usize,
) -> Result<IcaResult, String> {
    let n = x.rows;
    let d = x.cols;
    if n < 2 {
        return Err(format!("need at least 2 samples, got {}", n));
    }
    if x.data.iter().any(|v| !v.is_finite()) {
        return Err("X contains non-finite values (nan or inf)".to_string());
    }
    let k = n_components.unwrap_or(d);
    if k < 1 || k > d {
        return Err(format!(
            "n_components must be in [1, {}] for a {}x{} matrix, got {}",
            d, n, d, k
        ));
    }
    if max_iter < 1 {
        return Err(format!("max_iter must be a positive integer, got {}", max_iter));
    }

    // 1. Center.
    let mut mean = vec![0.0f64; d];
    for i in 0..n {
        for j in 0..d {
            mean[j] += x.get(i, j);
        }
    }
    for m in mean.iter_mut() {
        *m /= n as f64;
    }
    let mut xc = vec![0.0f64; n * d];
    for i in 0..n {
        for j in 0..d {
            xc[i * d + j] = x.get(i, j) - mean[j];
        }
    }

    // 2. Whiten. Cov = Xc^T Xc / (n - 1), d x d.
    let mut cov = vec![0.0f64; d * d];
    for a in 0..d {
        for b in a..d {
            let mut acc = 0.0;
            for i in 0..n {
                acc += xc[i * d + a] * xc[i * d + b];
            }
            let val = acc / (n as f64 - 1.0);
            cov[a * d + b] = val;
            cov[b * d + a] = val;
        }
    }
    let (eigvals, eigvecs) = jacobi_eigen(&cov, d);
    for &ev in eigvals.iter().take(k) {
        if ev <= 0.0 {
            return Err(
                "data is rank-deficient: a retained whitening direction has zero variance"
                    .to_string(),
            );
        }
    }
    // K = (E_k / sqrt(lambda_k))^T, shape k x d. E_k column c is eigvecs[*, c].
    let mut whitening = vec![vec![0.0f64; d]; k];
    for c in 0..k {
        let s = eigvals[c].sqrt();
        for i in 0..d {
            whitening[c][i] = eigvecs[i * d + c] / s;
        }
    }
    // Z = Xc K^T, n x k.
    let mut z = vec![vec![0.0f64; k]; n];
    for i in 0..n {
        for c in 0..k {
            let mut acc = 0.0;
            for j in 0..d {
                acc += xc[i * d + j] * whitening[c][j];
            }
            z[i][c] = acc;
        }
    }

    // 3. FastICA fixed-point iteration with symmetric orthogonalization.
    let mut identity = vec![vec![0.0f64; k]; k];
    for i in 0..k {
        identity[i][i] = 1.0;
    }
    let mut w = symmetric_decorrelation(&identity, k);
    let mut n_iter = 0usize;
    for _ in 0..max_iter {
        n_iter += 1;
        // WZ[i][j] = sum_t z[i][t] w[j][t]; gwz = tanh; g' = 1 - gwz^2.
        let mut w_new = vec![vec![0.0f64; k]; k];
        let mut gprime_mean = vec![0.0f64; k];
        // Accumulate E[z g(w^T z)] into w_new (k x k) and E[g'] into gprime_mean.
        for i in 0..n {
            for j in 0..k {
                let mut wz = 0.0;
                for t in 0..k {
                    wz += z[i][t] * w[j][t];
                }
                let g = wz.tanh();
                gprime_mean[j] += 1.0 - g * g;
                for t in 0..k {
                    w_new[j][t] += g * z[i][t];
                }
            }
        }
        for j in 0..k {
            gprime_mean[j] /= n as f64;
            for t in 0..k {
                w_new[j][t] = w_new[j][t] / n as f64 - gprime_mean[j] * w[j][t];
            }
        }
        w = symmetric_decorrelation(&w_new, k);
    }
    fix_signs(&mut w);

    // components = W K, k x d.
    let mut components = vec![vec![0.0f64; d]; k];
    for i in 0..k {
        for j in 0..d {
            let mut acc = 0.0;
            for c in 0..k {
                acc += w[i][c] * whitening[c][j];
            }
            components[i][j] = acc;
        }
    }

    // Order by descending recovered-source variance: S = Xc components^T, n x k.
    let mut var = vec![0.0f64; k];
    for c in 0..k {
        let mut sum = 0.0;
        let mut s_col = vec![0.0f64; n];
        for i in 0..n {
            let mut acc = 0.0;
            for j in 0..d {
                acc += xc[i * d + j] * components[c][j];
            }
            s_col[i] = acc;
            sum += acc;
        }
        let m = sum / n as f64;
        let mut ss = 0.0;
        for i in 0..n {
            let diff = s_col[i] - m;
            ss += diff * diff;
        }
        var[c] = ss / (n as f64 - 1.0);
    }
    let mut order: Vec<usize> = (0..k).collect();
    order.sort_by(|&a, &b| var[b].partial_cmp(&var[a]).unwrap());
    let w = order.iter().map(|&c| w[c].clone()).collect::<Vec<_>>();
    let components = order
        .iter()
        .map(|&c| components[c].clone())
        .collect::<Vec<_>>();

    let mixing = pseudoinverse(&components, k, d);

    Ok(IcaResult {
        mean,
        whitening,
        unmixing: w,
        components,
        mixing,
        n_iter,
    })
}

/// Recovers the independent sources from observed mixtures `x`.
///
/// Returns `n x n_components` scores: `S = (X - mean) components^T`.
pub fn transform(model: &IcaResult, x: &Matrix) -> Result<Vec<Vec<f64>>, String> {
    let d = model.n_features();
    if x.cols != d {
        return Err(format!(
            "X has {} features but model was fit on {}",
            x.cols, d
        ));
    }
    if x.data.iter().any(|v| !v.is_finite()) {
        return Err("X contains non-finite values (nan or inf)".to_string());
    }
    let k = model.n_components();
    let mut out = vec![vec![0.0f64; k]; x.rows];
    for i in 0..x.rows {
        for (c, comp) in model.components.iter().enumerate() {
            let mut acc = 0.0;
            for j in 0..d {
                acc += (x.get(i, j) - model.mean[j]) * comp[j];
            }
            out[i][c] = acc;
        }
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    // Shared fixtures: identical to the Python and Julia test suites.
    fn fixture() -> Matrix {
        let rows: Vec<Vec<f64>> = vec![
            vec![1.0, -2.0],
            vec![0.0, 5.0],
            vec![4.0, 2.0],
            vec![-2.0, -6.0],
            vec![-3.0, 1.0],
            vec![4.0, 7.0],
            vec![-3.0, -4.0],
            vec![1.0, 3.0],
        ];
        Matrix::from_rows(&rows).unwrap()
    }

    const TOL: f64 = 1e-9;

    #[test]
    fn mean_matches_fixture() {
        let r = fit_ica(&fixture(), Some(2), 200).unwrap();
        let expected: [f64; 2] = [0.25, 0.75];
        assert!((r.mean[0] - expected[0]).abs() < TOL);
        assert!((r.mean[1] - expected[1]).abs() < TOL);
    }

    // ICA recovers components only up to sign and permutation, and a hand-rolled
    // Jacobi eigensolver can choose different signs/order than numpy's `eigh`.
    // We therefore match the numpy-derived fixtures up to per-row sign and row
    // permutation, which are the genuine ICA invariants. (The Python and Julia
    // suites pin the same fixtures; their solvers happen to land on this exact
    // sign convention, Rust's lands on an equivalent one.)
    fn rows_match_up_to_sign_perm(got: &[Vec<f64>], exp: &[[f64; 2]; 2]) -> bool {
        let row_eq = |a: &[f64], b: &[f64]| {
            let same = (a[0] - b[0]).abs() < TOL && (a[1] - b[1]).abs() < TOL;
            let neg = (a[0] + b[0]).abs() < TOL && (a[1] + b[1]).abs() < TOL;
            same || neg
        };
        (row_eq(&got[0], &exp[0]) && row_eq(&got[1], &exp[1]))
            || (row_eq(&got[0], &exp[1]) && row_eq(&got[1], &exp[0]))
    }

    #[test]
    fn unmixing_matches_fixture() {
        let r = fit_ica(&fixture(), Some(2), 200).unwrap();
        let expected: [[f64; 2]; 2] = [
            [0.7846180154937991, 0.6199794914048147],
            [-0.6199794914048147, 0.784618015493799],
        ];
        assert!(rows_match_up_to_sign_perm(&r.unmixing, &expected));
    }

    #[test]
    fn components_match_fixture() {
        let r = fit_ica(&fixture(), Some(2), 200).unwrap();
        let expected: [[f64; 2]; 2] = [
            [0.3534839303856012, 0.0016237326826006237],
            [0.2994403038579343, -0.2922019487902003],
        ];
        assert!(rows_match_up_to_sign_perm(&r.components, &expected));
    }

    #[test]
    fn transform_first_row_matches_fixture() {
        let r = fit_ica(&fixture(), Some(2), 200).unwrap();
        let s = transform(&r, &fixture()).unwrap();
        let expected: [f64; 2] = [0.2606476829120492, 1.0281355870665014];
        // up to component permutation and sign: compare sorted absolute values
        let mut got = [s[0][0].abs(), s[0][1].abs()];
        let mut exp = [expected[0].abs(), expected[1].abs()];
        got.sort_by(|a, b| a.partial_cmp(b).unwrap());
        exp.sort_by(|a, b| a.partial_cmp(b).unwrap());
        assert!((got[0] - exp[0]).abs() < TOL && (got[1] - exp[1]).abs() < TOL);
    }

    #[test]
    fn unmixing_is_orthogonal() {
        let r = fit_ica(&fixture(), Some(2), 200).unwrap();
        let w = &r.unmixing;
        // W W^T = I.
        for i in 0..2 {
            for j in 0..2 {
                let mut acc = 0.0;
                for t in 0..2 {
                    acc += w[i][t] * w[j][t];
                }
                let expected = if i == j { 1.0 } else { 0.0 };
                assert!((acc - expected).abs() < TOL);
            }
        }
    }

    #[test]
    fn mixing_inverts_components() {
        let r = fit_ica(&fixture(), Some(2), 200).unwrap();
        // components @ mixing = I.
        for i in 0..2 {
            for j in 0..2 {
                let mut acc = 0.0;
                for t in 0..2 {
                    acc += r.components[i][t] * r.mixing[t][j];
                }
                let expected = if i == j { 1.0 } else { 0.0 };
                assert!((acc - expected).abs() < TOL);
            }
        }
    }

    #[test]
    fn n_iter_equals_max_iter() {
        let r = fit_ica(&fixture(), Some(2), 37).unwrap();
        assert_eq!(r.n_iter, 37);
    }

    #[test]
    fn too_few_samples_errors() {
        let x = Matrix::from_rows(&[vec![1.0, 2.0]]).unwrap();
        assert!(fit_ica(&x, None, 200).is_err());
    }

    #[test]
    fn bad_n_components_errors() {
        assert!(fit_ica(&fixture(), Some(5), 200).is_err());
    }

    #[test]
    fn bad_max_iter_errors() {
        assert!(fit_ica(&fixture(), Some(2), 0).is_err());
    }
}
